clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1746,6 +1746,87 @@ segmented_reduce_argmax(Policy &&policy, Iter1 keys_in, Iter2 keys_out,
   policy.queue().wait();
 }
 
+template <typename ExecutionPolicy, typename InputIterator,
+          typename OutputIterator1, typename OutputIterator2,
+          typename OutputIterator3>
+::std::pair<OutputIterator1, OutputIterator2> nontrivial_run_length_encode(
+    ExecutionPolicy &&policy, InputIterator input_beg, InputIterator input_end,
+    OutputIterator1 offsets_out, OutputIterator2 lengths_out,
+    OutputIterator3 num_runs) {
+  using oneapi::dpl::make_transform_iterator;
+  using oneapi::dpl::make_zip_iterator;
+  using offsets_t =
+      typename ::std::iterator_traits<OutputIterator1>::value_type;
+  using lengths_t =
+      typename ::std::iterator_traits<OutputIterator2>::value_type;
+
+  auto n = ::std::distance(input_beg, input_end);
+  // First element must be nontrivial run (start of first segment)
+  auto first_adj_it = oneapi::dpl::adjacent_find(policy, input_beg, input_end);
+  auto first_adj_idx = ::std::distance(input_beg, first_adj_it);
+  if (first_adj_it == input_end) {
+    ::std::fill(policy, num_runs, num_runs + 1, 0);
+    return ::std::make_pair(offsets_out, lengths_out);
+  }
+  auto get_prev_idx_element = [first_adj_idx](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == 0) ? 0 : out_idx - 1;
+  };
+  auto get_next_idx_element = [first_adj_idx, n](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == n - 1) ? n - 1 : out_idx + 1;
+  };
+  // TODO: Use shifted view to pad range once oneDPL ranges is non-experimental
+  auto left_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_prev_idx_element);
+  auto right_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_next_idx_element);
+  // Segment type for ith idx consists of zip of iterators at (i-1, i, i+1)
+  // padded at the ends
+  auto zipped_keys_beg = make_zip_iterator(
+      left_shifted_input_beg, input_beg, right_shifted_input_beg,
+      oneapi::dpl::counting_iterator<offsets_t>(0));
+  // Set flag at the beginning of new nontrivial run (ex: (0, 1, 1) -> 1)
+  auto key_flags_beg =
+      make_transform_iterator(zipped_keys_beg, [n](const auto &zipped) {
+        using ::std::get;
+        bool zero_idx_mask = get<3>(zipped) != 0;
+        bool last_idx_mask = get<3>(zipped) != n - 1;
+        return (zero_idx_mask * (get<0>(zipped) != get<1>(zipped)) &&
+                get<1>(zipped) == get<2>(zipped)) *
+               last_idx_mask;
+      });
+  auto count_beg = oneapi::dpl::counting_iterator<offsets_t>(0);
+  auto const_it = dpct::make_constant_iterator(lengths_t(1));
+  // Check for presence of nontrivial element at current index
+  auto tr_nontrivial_flags = make_transform_iterator(
+      make_zip_iterator(left_shifted_input_beg, input_beg,
+                        right_shifted_input_beg),
+      [](const auto &zip) {
+        using ::std::get;
+        return get<0>(zip) == get<1>(zip);
+      });
```

Comment:
Do we need the `(i+1)` input here in the zip?  The zip only needs 2 elements, since that is all the transform uses.  

The fact that we seed the count at 1 is good, and allows us to make use the of knowledge a non-trivial run exists in this segment and avoid a second comparison for each element.

Generally, I think it ends up being equivalent to compare `*i == *(i+1)` vs comparing `*(i-1) == *i`.  However, depending on which we choose, I think we need special casing for the first or the last element, due to the padding strategy of repeating the element.  Without it I believe we can end up with an inflated result (+1) for our first or last non-trivial run.

Reply:
We do not need the (i+1) input here. I have removed `right_shifted_input_beg`, so we end up just comparing `*(i-1) == *i`.

Also, there is no correctness issue in this case. The bogus `1` flag set when we compare `i-1` and `i` when `i=0` never affects the run count as it never appears as a `rhs` argument in `op`. I have added a comment to explain this. It makes the code a bit more complicated but allows us to have less comparisons in our operator and start the count at 1. This case is also added in the test PR which passes.